addons/pos_coupon/static/src/js/PaymentScreen.js

```diff
@@ -8,27 +8,29 @@ odoo.define('pos_coupon.PaymentScreen', function (require) {
     const PosCouponPaymentScreen = (PaymentScreen) =>
         class extends PaymentScreen {
             async _postPushOrderResolve(order, server_ids) {
-                const bookedCouponIds = new Set(
-                    Object.values(order.bookedCouponCodes)
-                        .map((couponCode) => couponCode.coupon_id)
-                        .filter((coupon_id) => coupon_id)
-                );
-                const usedCouponIds = order.orderlines.models
-                    .map((line) => line.coupon_id)
-                    .filter((coupon_id) => coupon_id);
-                for (let coupon_id of usedCouponIds) {
-                    bookedCouponIds.delete(coupon_id);
+                if(this.env.pos.coupon_programs.length) {
+                    const bookedCouponIds = new Set(
+                        Object.values(order.bookedCouponCodes)
+                            .map((couponCode) => couponCode.coupon_id)
+                            .filter((coupon_id) => coupon_id)
+                    );
+                    const usedCouponIds = order.orderlines.models
+                        .map((line) => line.coupon_id)
+                        .filter((coupon_id) => coupon_id);
+                    for (let coupon_id of usedCouponIds) {
+                        bookedCouponIds.delete(coupon_id);
+                    }
+                    const unusedCouponIds = [...bookedCouponIds.values()];
+                    order.generated_coupons = await this.rpc(
+                        {
+                            model: 'pos.order',
+                            method: 'validate_coupon_programs',
+                            args: [server_ids, order.programIdsToGenerateCoupons || [], unusedCouponIds],
+                            kwargs: { context: session.user_context },
+                        },
+                        {}
+                    );
```

Comment:
To reduce the diff can you instead do something like this:

```suggestion
                if (this.env.pos.coupon_programs.length === 0) {
                    return super._postPushOrderResolve(order, server_ids);
                }
```

Basically, we immediately call the the super if there are no coupon programs.